glib-macros/src/boxed_derive.rs

```diff
@@ -197,6 +197,22 @@ pub fn impl_boxed(input: &syn::DeriveInput) -> TokenStream {
 
         #impl_from_value
 
+        impl #crate_ident::translate::FromGlibPtrBorrow<*const #name> for #name {
+            #[inline]
+            unsafe fn from_glib_borrow(ptr: *const #name) -> #crate_ident::translate::Borrowed<Self> {
+                #crate_ident::translate::FromGlibPtrBorrow::from_glib_borrow(ptr as *mut _)
+            }
+        }
+
+        impl #crate_ident::translate::FromGlibPtrBorrow<*mut #name> for #name {
+            #[inline]
+            unsafe fn from_glib_borrow(ptr: *mut #name) -> #crate_ident::translate::Borrowed<Self> {
+                debug_assert!(!ptr.is_null());
+
+                #crate_ident::translate::Borrowed::new(*::std::boxed::Box::from_raw(ptr))
```

Comment:
This is stealing the memory :) The `Borrowed` is not going to free it later so it's technically fine, but going via a cast and dereference or so would be better

Reply:
Yeah, I was indeed relying on the `ManuallyDrop` inside the `Borrowed` to deal with that.
The problem is that I'm not 100% sure we can cast the raw pointer into the `Box` in an easy way.

The `as` keyword will work only for primitive types, maybe we could use `mem::transmute` but it's not going to be pretty either.

And actually, looking at the `from_raw` implementation of `Box`, I guess we would be missing the allocator part of the `Box` in the conversion:
https://doc.rust-lang.org/src/alloc/boxed.rs.html#997-999

The same thing for the shared type actually, which is also stealing the ptr with the `from_raw` implementation of the `RefCounted` trait.

Any specific idea on how to fix that?

Reply:
Also dereferencing directly the raw pointer without going through the `Box` is not going to work:

```
#crate_ident::translate::Borrowed::new(*ptr)
```

```
error[E0507]: cannot move out of `*ptr` which is behind a raw pointer
```

Reply:
Hrm I see, we need a full `T` here. I think this is the only way then... I don't like it but OK :)
Can you add a comment why the `from_raw()` is not breaking everything here?

Reply:
sure, just added that :)

Reply:
Wait, you're deconstructing the box here with the dereference. That's actually not valid like this. You would free the heap memory of the box and move out the `T` from there onto the stack.

Reply:
Yes you're right, my apologies.

I will need to find another way, even if there's nothing obvious that comes to my mind.
One possibility maybe would be wrapping the box in a `ManuallyDrop` before dereferencing it, I can maybe try that out

Reply:
`ptr::read()` from the `*const T` should work as long as it's wrapped in the `ManuallyDrop` / `Borrowed` right after